### PR TITLE
xds: allow updating listener back to original state

### PR DESCRIPTION
**Pull request by samflattery**
Signed-off-by: Sam Flattery <samflattery@google.com>

Commit Message: Allow updating listener back to original state
Additional Description:

- since `addOrUpdateListenerInternal` returns early in the case of a duplicate active/warming listener being added, it means you cannot update a listener back to its original state after updating it to a warming state
- consider the following sequence of actions: 
1. add `listener_0` referencing `route_config_0`, make it active
2. update `listener_0` referencing `route_config_1`, keep it warming (Envoy keeps the original `listener_0` active until the new warms)
3. update `listener_0` back to `route_config_0`, which should remove the warming listener and return Envoy to its initial state, so that a future addition of `route_config_1` won't cause the listener to change to that state, but right now it doesn't
- make a small change in `listener_manager_impl.cc` to allow this to happen
- update `xds_verifier.cc` and add a unit test for this case

I can also add a unit test to `ads_integration_test.cc` if required

- this issue was found by my xDS fuzzer
- added regression corpus

Risk Level: Medium
Testing: Passes existing fuzz corpora and a few minutes on libfuzzer, passes `ads_integration_test` and `lds_api_test`
Docs Changes: N/A
Release Notes: N/A
Fixes: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=24888

### CODE REVIEW

File: source/server/listener_manager_impl.cc [hunk 1]
```diff
@@ -393,6 +393,17 @@ bool ListenerManagerImpl::addOrUpdateListenerInternal(
   auto existing_active_listener = getListenerByName(active_listeners_, name);
   auto existing_warming_listener = getListenerByName(warming_listeners_, name);
 
+  // The listener should be updated back to its original state and the warming listener should be
```
Comment (htuch) [inline source/server/listener_manager_impl.cc:396]: Can you add a `listener_manager_impl_test` unit test for this?

**Comment by lambdai** [comment]
Removing a warming listener is designed to be allowed. So this PR should help Envoy moving to the desired state faster.
What's more this PR allows for A-warmingB-A update.

Great!

File: test/server/config_validation/xds_verifier.cc [hunk 2]
```diff
@@ -81,10 +81,37 @@ void XdsVerifier::listenerUpdated(const envoy::config::listener::v3::Listener& l
...
-      })) {
+               getRoute(rep.listener) == getRoute(listener) && rep.state == WARMING;
+      });
+  const auto existing_active =
+      std::find_if(listeners_.begin(), listeners_.end(), [&](const auto& rep) {
+        return rep.listener.name() == listener.name() &&
+               getRoute(rep.listener) == getRoute(listener) && rep.state == ACTIVE;
+      });
+
+  // return early if the listener is a duplicate
```
Comment (lambdai) [inline test/server/config_validation/xds_verifier.cc:95]: Capitalize the first letter before you are rejected by Matt :)

File: source/server/listener_manager_impl.cc [hunk 3]
```diff
@@ -393,6 +393,17 @@ bool ListenerManagerImpl::addOrUpdateListenerInternal(
...
+  // removed.
+  if (existing_warming_listener != warming_listeners_.end() &&
+      existing_active_listener != active_listeners_.end() &&
+      (*existing_active_listener)->blockUpdate(hash)) {
```
Comment (lambdai) [inline source/server/listener_manager_impl.cc:400]: I have to look into blockUpdate before I asked how to verify the active listener and the latest config is equivalent. Sigh...

Comment (samflattery) [reply source/server/listener_manager_impl.cc:400]: `blockUpdate()` just checks that the hash of the two listeners is the same
https://github.com/envoyproxy/envoy/blob/51551ae944c642e6fc61563cbea8653087e70f1f/source/server/listener_impl.h#L256

https://github.com/envoyproxy/envoy/blob/51551ae944c642e6fc61563cbea8653087e70f1f/source/common/protobuf/utility.cc#L264-L277

It's used just below this snippet to ignore duplicate listeners: https://github.com/envoyproxy/envoy/blob/25fcc8b40523d74c28495208cae2cc4cf701d98b/source/server/listener_manager_impl.cc#L409-L415

File: test/server/config_validation/xds_verifier.cc (hunk 2 above)
Comment (samflattery) [reply test/server/config_validation/xds_verifier.cc:95]: All of the comments in `xds_verifier.cc` have a small first letter, I'll go through and change them all

File: source/server/listener_manager_impl.cc (hunk 1 above)
Comment (samflattery) [reply source/server/listener_manager_impl.cc:396]: added!

File: test/server/config_validation/xds_verifier_test.cc [hunk 4]
```diff
@@ -179,49 +179,68 @@ TEST(XdsVerifier, WarmMultipleListenersDelta) {
...
+  // Send a route that will be ignored.
   verifier.routeAdded(buildRoute("route_config_0"));
 
-  // add a warming listener that refers to this route
+  // Add a warming listener that refers to this route.
   verifier.listenerAdded(buildListener("listener_0", "route_config_0"));
   EXPECT_TRUE(verifier.hasListener("listener_0", verifier.WARMING));
 
-  // send the same route again, make sure listener becomes active
+  // Send the same route again, make sure listener becomes active.
```
Comment (htuch) [inline test/server/config_validation/xds_verifier_test.cc:204]: Nit: please split out these bulk formatting changes to distinct PRs in future, as they make it hard to review the substantive changes.

**Pull request merged**